src/vsr/replica_test.zig

```diff
@@ -0,0 +1,920 @@
+    try expectEqual(t.replica(.R_).commit(), 20);
+
+    t.replica(.R0).stop();
+    try expectEqual(c.request(24), 24);
+    t.replica(.R1).stop();
+    t.replica(.R2).stop();
+
+    t.replica(.R1).corrupt(.{ .wal_prepare = 22 });
+
+    // The nack quorum size is 2.
+    // The new view must determine whether op=22 is possibly committed.
+    // - R0 never received op=22 (it had already crashed), so it nacks.
+    // - R1 did receive op=22, but upon recovering its WAL, it was corrupt, so it cannot nack.
+    // The cluster must wait form R2 before recovering.
+    try expectEqual(t.replica(.R0).open(), .ok);
+    try expectEqual(t.replica(.R1).open(), .ok);
+    // TODO Explicit code coverage marks: This should hit the "quorum received, awaiting repair"
+    // log line in on_do_view_change().
+    t.run();
+    try expectEqual(t.replica(.R0).status(), .view_change);
+    try expectEqual(t.replica(.R1).status(), .view_change);
+
+    // R2 provides the missing header, allowing the view-change to succeed.
+    try expectEqual(t.replica(.R2).open(), .ok);
+    t.run();
+    try expectEqual(t.replica(.R_).status(), .normal);
+    try expectEqual(t.replica(.R_).commit(), 24);
+}
+
+test "Cluster: view-change: DVC, 2/3 faulty header stall" {
+    const t = try TestContext.init(.{ .replica_count = 3 });
+    defer t.deinit();
+
+    var c = t.clients(0, t.cluster.clients.len);
+    try expectEqual(c.request(20), 20);
+    try expectEqual(t.replica(.R_).commit(), 20);
+
+    t.replica(.R0).stop();
+    try expectEqual(c.request(24), 24);
+    t.replica(.R1).stop();
+    t.replica(.R2).stop();
+
+    t.replica(.R1).corrupt(.{ .wal_prepare = 22 });
+    t.replica(.R2).corrupt(.{ .wal_prepare = 22 });
+
+    try expectEqual(t.replica(.R_).open(), .ok);
+    // TODO Explicit code coverage marks: This should hit the "quorum received, deadlocked"
+    // log line in on_do_view_change().
+    t.run();
+    try expectEqual(t.replica(.R_).status(), .view_change);
+}
+
+const ProcessSelector = enum {
+    __, // all replicas, standbys, and clients
+    R_, // all (non-standby) replicas
+    R0,
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    S_, // all standbys
+    S0,
+    S1,
+    S2,
+    S3,
+    S4,
+    S5,
+    A0, // current primary
+    B1, // backup immediately following current primary
+    B2,
+    B3,
+    B4,
+    B5,
+    C_, // all clients
+};
+
+const TestContext = struct {
+    cluster: *Cluster,
+    log_level: std.log.Level,
+    client_requests: [constants.clients_max]usize = [_]usize{0} ** constants.clients_max,
+    client_replies: [constants.clients_max]usize = [_]usize{0} ** constants.clients_max,
+
+    pub fn init(options: struct {
+        replica_count: u8,
+        standby_count: u8 = 0,
+        client_count: u8 = constants.clients_max,
+    }) !*TestContext {
+        var log_level_original = std.testing.log_level;
+        std.testing.log_level = log_level;
+
+        var prng = std.rand.DefaultPrng.init(123);
```

Comment:
:thinking: I am wondering if this should be genuinely random, to make sure the tests are deterministic for real, and not just accidetally?

Reply:
Maybe with a seed from an env variable? I don't actually want to get a random number from the OS here since we want CI to be deterministic (w/ a consistent seed).